test/integration/job/job_test.go

```diff
@@ -494,6 +494,374 @@ func TestJobPodFailurePolicy(t *testing.T) {
+		wantFailed           int
+		wantSucceeded        int
+		wantActiveIndexes    sets.Set[int]
+		wantCompletedIndexes string
+		wantFailedIndexes    *string
+	}
+
+	podTemplateSpec := v1.PodTemplateSpec{
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:                     "main-container",
+					Image:                    "foo",
+					ImagePullPolicy:          v1.PullIfNotPresent,
+					TerminationMessagePolicy: v1.TerminationMessageFallbackToLogsOnError,
+				},
+			},
+		},
+	}
+	testCases := map[string]struct {
+		enableJobSuccessPolicy     bool
+		enableBackoffLimitPerIndex bool
+		job                        batchv1.Job
+		podTerminations            []podTerminationWithExpectations
+		wantConditionTypes         []batchv1.JobConditionType
+		wantJobFinishedNumMetric   []metricLabelsWithValue
+	}{
+		"all indexes succeeded; JobSuccessPolicy is enabled": {
+			enableJobSuccessPolicy: true,
+			job: batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism:    ptr.To[int32](1),
+					Completions:    ptr.To[int32](1),
+					CompletionMode: completionModePtr(batchv1.IndexedCompletion),
+					Template:       podTemplateSpec,
+					SuccessPolicy: &batchv1.SuccessPolicy{
+						Criteria: []batchv1.SuccessPolicyCriteria{{
+							SucceededIndexes: ptr.To("0"),
+						}},
+					},
+				},
+			},
+			podTerminations: []podTerminationWithExpectations{
+				{
+					index: 0,
+					status: v1.PodStatus{
+						Phase: v1.PodSucceeded,
+					},
+					wantActive:           0,
+					wantFailed:           0,
+					wantSucceeded:        1,
+					wantCompletedIndexes: "0",
+				},
+			},
+			wantConditionTypes: []batchv1.JobConditionType{batchv1.JobSuccessCriteriaMet, batchv1.JobComplete},
+			wantJobFinishedNumMetric: []metricLabelsWithValue{
+				{
+					Labels: []string{"Indexed", "succeeded", ""},
+					Value:  1,
+				},
+			},
+		},
+		"all indexes succeeded; JobSuccessPolicy is disabled": {
+			job: batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism:    ptr.To[int32](1),
+					Completions:    ptr.To[int32](1),
+					CompletionMode: completionModePtr(batchv1.IndexedCompletion),
+					Template:       podTemplateSpec,
+					SuccessPolicy: &batchv1.SuccessPolicy{
+						Criteria: []batchv1.SuccessPolicyCriteria{{
+							SucceededIndexes: ptr.To("0"),
+						}},
+					},
+				},
+			},
+			podTerminations: []podTerminationWithExpectations{
+				{
+					index: 0,
+					status: v1.PodStatus{
+						Phase: v1.PodSucceeded,
+					},
+					wantActive:           0,
+					wantFailed:           0,
+					wantSucceeded:        1,
+					wantCompletedIndexes: "0",
+				},
+			},
+			wantConditionTypes: []batchv1.JobConditionType{batchv1.JobComplete},
+			wantJobFinishedNumMetric: []metricLabelsWithValue{
+				{
+					Labels: []string{"Indexed", "succeeded", ""},
+					Value:  1,
+				},
+			},
+		},
+		"job with successPolicy with succeededIndexes; job has SuccessCriteriaMet and Complete conditions even if some indexes remain pending": {
```

Comment:
Shouldn't we wait for Pods to be Failed or Succeeded before declaring the job Complete?

Reply:
That makes sense. To verify the situation, we need to separate the Test function.
I'll try to add it.

Reply:
Done.